pkg/apiserver/rest/usecase/application.go

```
@@ -1415,3 +1493,220 @@ func genWebhookToken() string {
 	}
 	return string(b)
 }
+
+func convertToModelComponent(appPrimaryKey string, component common.ApplicationComponent) model.ApplicationComponent {
+	bc := model.ApplicationComponent{
+		AppPrimaryKey:    appPrimaryKey,
+		Name:             component.Name,
+		Type:             component.Type,
+		ExternalRevision: component.ExternalRevision,
+		DependsOn:        component.DependsOn,
+		Inputs:           component.Inputs,
+		Outputs:          component.Outputs,
+		Scopes:           component.Scopes,
+	}
+	if component.Properties != nil {
+		bc.Properties, _ = model.NewJSONStruct(component.Properties)
+	}
+	for _, trait := range component.Traits {
+		properties, _ := model.NewJSONStruct(trait.Properties)
```

we should report the error if json marshal failed